docs: add ESM section to avoid configuration issues (#1782)
libp2p/js-libp2p: doc/GETTING_STARTED.md

If you're new to libp2p, we recommend configuring your node in stages, as this can make troubleshooting configuration issues much easier. In this guide, we'll do just that. If you're more experienced with libp2p, you may wish to jump to the [Configuration readme](./CONFIGURATION.md).

> ## ESM

Reviewer: ```suggestion
### ESM
```

this needs to be the same level as it’s sibling headers, otherwise “Basic Setup” etc become sub-sections of “ESM”

the TOC will also need updating.

> Since `libp2p@0.37.0` is that the module is now [ESM-only](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c).

Reviewer: ```suggestion
Since `libp2p@0.37.0` the module is now [ESM-only](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c).
```

ESM is the module system for JavaScript, it allows us to structure our code in separate files without polluting a global namespace.

Other systems have tried to fill this gap, notably CommonJS, AMD, RequireJS and others, but ESM is [the official standard format](https://tc39.es/ecma262/#sec-modules) to package JavaScript code for reuse.

> 

Reviewer: I think it’s worth adding a section with the minimal typescript config for generating ESM here too

Author: Well someone could use ESM modules without Typescript but I suppose we suggest that  the `type` parameter be set to `module` as mentioned in the [node docs](https://nodejs.org/api/esm.html#enabling)

Reviewer: That is of course true, but by default TypeScript outputs CJS which then explodes at runtime.

People are then confused because since TypeScript uses `import` it looks like they are authoring ESM when in fact they are not.

A cursory search through the closed issues in js-IPFS around the time we converted to ESM will reveal that this is a very common mistake.

If you're new to libp2p, we recommend configuring your node in stages, as this can make troubleshooting configuration issues much easier. In this guide, we'll do just that. If you're more experienced with libp2p, you may wish to jump to the [Configuration readme](./CONFIGURATION.md).

### ESM

Since `libp2p@0.37.0` modules are now [ESM-only](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c).

ESM is the module system for JavaScript, it allows us to structure our code in separate files without polluting a global namespace.

Other systems have tried to fill this gap, notably CommonJS, AMD, RequireJS and others, but ESM is [the official standard format](https://tc39.es/ecma262/#sec-modules) to package JavaScript code for reuse. This means that you need ensure your configuration uses the correct module system, if you are using Typescript, set the [`module` field in your tsconfig](https://www.typescriptlang.org/tsconfig#module) to `ES2022 ` or later e.g.

```json
{
  "compilerOptions": {
    "module": "ES2022",
    "esModuleInterop": true,
    "target": "ES2022",
    "moduleResolution": "node"
  }
}

> For more info on enablng ES modules in Node, see [this guide](https://nodejs.org/api/esm.html).
> ### Basic setup

Reviewer: ```suggestion
For more info on enablng ES modules in Node, see [this guide](https://nodejs.org/api/esm.html).

### Basic setup
```

If you're new to libp2p, we recommend configuring your node in stages, as this can make troubleshooting configuration issues much easier. In this guide, we'll do just that. If you're more experienced with libp2p, you may wish to jump to the [Configuration readme](./CONFIGURATION.md).

### ESM

Since `libp2p@0.37.0` modules are now [ESM-only](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c).

ESM is the module system for JavaScript, it allows us to structure our code in separate files without polluting a global namespace.

> Other systems have tried to fill this gap, notably CommonJS, AMD, RequireJS and others, but ESM is [the official standard format](https://tc39.es/ecma262/#sec-modules) to package JavaScript code for reuse. This means that you need ensure your configuration uses the correct module system, if you are using Typescript, set the [`module` field in your tsconfig](https://www.typescriptlang.org/tsconfig#module) to your desired ESM version e.g.

Reviewer: ```suggestion
Other systems have tried to fill this gap, notably CommonJS, AMD, RequireJS and others, but ESM is [the official standard format](https://tc39.es/ecma262/#sec-modules) to package JavaScript code for reuse. This means that you need ensure your configuration uses the correct module system, if you are using Typescript, set the [`module` field in your tsconfig](https://www.typescriptlang.org/tsconfig#module) to `ES2022 ` or later e.g.
```

```json
{
  "compilerOptions": {
    "module": "ES2022",
    "esModuleInterop": true,
    "target": "ES2022",
    "moduleResolution": "node"
  }
}

> ```
> For more info on enablng ES modules in Node, see [this guide](https://nodejs.org/api/esm.html).

Reviewer: ```suggestion
```

For more info on enablng ES modules in Node, see [this guide](https://nodejs.org/api/esm.html).
```
